File: frontend/src/api/screenshot.ts
/**
 * `screenshot.*` — every instance's screenshots as one listing. Nothing is
 * copied: a shot is read where the game wrote it, so the shell has to let the
 * webview into that folder before its `asset:` URL resolves.
 */
import { convertFileSrc } from '@tauri-apps/api/core';

import { call, invokeCommand } from './core/ipc';
import type { Screenshot } from './types/screenshot';

export async function list(instance = ''): Promise<Screenshot[]> {
  const result = await call<{ screenshots: Screenshot[] }>('screenshot.list', {
    instance,
  });
  return result.screenshots;
}

export async function remove(
  instance: string,
  file: string,
): Promise<Screenshot[]> {
  const result = await call<{ screenshots: Screenshot[] }>(
    'screenshot.delete',
    { instance, file },
  );
  return result.screenshots;
}

export function allow(dirs: string[]): Promise<void> {
  return invokeCommand('screenshots_allow', { dirs });
}

export function folders(screenshots: Screenshot[]): string[] {
  const dirs = screenshots.map((shot) =>
    shot.path.slice(
      0,
      Math.max(shot.path.lastIndexOf('/'), shot.path.lastIndexOf('\\')),
    ),
  );
  return [...new Set(dirs.filter((dir) => dir.length > 0))];
}

export function url(shot: Screenshot): string {
  return `${convertFileSrc(shot.path)}?v=${shot.takenUnix}`;
}

File: crates/client/src/facades/screenshot.rs
//! `screenshot.*` — every instance's screenshots as one listing.

use proto::screenshot::{
    Screenshot as Shot, ScreenshotDelete, ScreenshotDeleteParams, ScreenshotList,
    ScreenshotListParams,
};

use crate::{IpcError, Session};

pub struct Screenshot<'a> {
    pub(crate) session: &'a Session,
}

impl Screenshot<'_> {
    /// Every instance that shares them; one instance when named.
    pub async fn list(&self, instance: &str) -> Result<Vec<Shot>, IpcError> {
        Ok(self
            .session
            .call::<ScreenshotList>(&ScreenshotListParams {
                instance: instance.to_string(),
            })
            .await?
            .screenshots)
    }

    pub async fn delete(&self, instance: &str, file: &str) -> Result<Vec<Shot>, IpcError> {
        Ok(self
            .session
            .call::<ScreenshotDelete>(&ScreenshotDeleteParams {
                instance: instance.to_string(),
                file: file.to_string(),
            })
            .await?
            .screenshots)
    }
}
